b2sdk/stream/chained.py

```diff
@@ -0,0 +1,129 @@
+    when chanined stream is closed.
+    """
+
+    def __init__(self, stream_openers):
+        """
+        :param list stream_openeres: list of callables that return opened steam
+        """
+        stream_openers = list(stream_openers)
+        if not stream_openers:
+            raise ValueError('chain_links cannot be empty')
+        self.stream_openers = stream_openers
+        self._stream_openers_iterator = iter(self.stream_openers)
+        self._current_stream = None
+        self._pos = 0
+
+    @property
+    def stream(self):
+        """ Return currently processed stream. """
+        if self._current_stream is None:
+            self._next_stream()
+        return self._current_stream
+
+    def _reset_chain(self):
+        if self._current_stream is not None:
+            self._current_stream.close()
+            self._current_stream = None
+        self._stream_openers_iterator = iter(self.stream_openers)
+        self._pos = 0
+
+    def _next_stream(self):
+        next_stream_opener = next(self._stream_openers_iterator, None)
+        if next_stream_opener is not None:
+            if self._current_stream is not None:
+                self._current_stream.close()
+            self._current_stream = next_stream_opener()
+
+    def seekable(self):
+        return True
+
+    def tell(self):
+        return self._pos
+
+    def seek(self, pos, whence=0):
+        """
+        Resets stream to the beginning.
+
+        :param int pos: only allowed value is ``0``
+        :param int whence: only allowed value is ``0``
+        """
+        if pos != 0 or whence != 0:
+            raise io.UnsupportedOperation('Chained stream can only be seeked to beginning')
+
+        self._reset_chain()
+
+        return self.tell()
+
+    def readable(self):
+        return True
+
+    def read(self, size=None):
+        """
+        Read data from the stream.
+
+        :param int,None size: number of bytes to read. If omitted, ``None``,
+                    or negative data is read and returned until EOF is reached
+
+        :return: data read from the stream
+        """
+        if size < 0:
+            size = None
+        byte_arrays = []
+        remaining = size
+        while 1:
+            current_stream = self.stream
+            buff = current_stream.read(remaining)
+            byte_arrays.append(buff)
+            if remaining is None or not buff:
+                self._next_stream()
+                if self.stream is current_stream:
+                    break
+            else:
+                remaining -= len(buff)
+
+        if not byte_arrays:
+            data = byte_arrays[0]
+        else:
+            data = bytes().join(byte_arrays)
+        self._pos += len(data)
+        return data
+
+    def close(self):
+        if self._current_stream is not None:
+            self._current_stream.close()
+        for stream_opener in self.stream_openers:
+            if isinstance(stream_opener, StreamOpener):
```

Comment:
could we instead check if stream opener has a cleanup method and call that or just try to call it and catch attributeerror? In Python someone can pass a duck and they don't need to inherit from `StreamOpener`